# Node omits configured skips from TAP. Report the classifier's decisions
# to distinguish deliberate exclusions from missing results.
# https://github.com/nodejs/node/blob/d57af10d1/tools/test.py#L1763-L1776
--- a/tools/test.py
+++ b/tools/test.py
@@ -30,6 +30,7 @@
 
 from __future__ import print_function
 from typing import Dict
+import json
 import logging
 import optparse
 import os
@@ -1762,13 +1763,19 @@
 
   def should_keep(case):
     if any((s in case.file) for s in options.skip_tests):
-      return False
+      reason = '--skip-tests'
     elif SKIP in case.outcomes:
-      return False
+      reason = 'status-file SKIP'
     elif (options.flaky_tests == SKIP) and (set([SLOW, FLAKY]) & case.outcomes):
-      return False
+      reason = '--flaky-tests=skip'
     else:
       return True
+    if options.progress == 'tap':
+      logger.info('# gvisor-node-skip %s', json.dumps({
+        'name': NormalizePath(case.file, test_root + os.sep),
+        'reason': reason,
+      }))
+    return False
 
   cases_to_run = [
     test_case for test_case in all_cases if should_keep(test_case)
